dbms/src/Operators/Operator.h

```diff
@@ -0,0 +1,133 @@
+// Copyright 2023 PingCAP, Ltd.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <Core/Block.h>
+
+#include <memory>
+
+namespace DB
+{
+/**
+ * All interfaces of the operator may return the following state.
+ * - finish status, spilling status and waiting status can be returned in all interfaces of operator.
+ * - operator may return a different running status depending on the interface.
+*/
+enum class OperatorStatus
+{
+    /// finish status
+    FINISHED,
+    CANCELLED,
+    /// spilling status
+    SPILLING,
+    /// waiting status
+    WAITING,
+    /// running status
+    // means that TransformOp/SinkOp needs to input a block to do the calculation,
+    NEED_INPUT,
+    // means that SourceOp/TransformOp outputs a block as input to the subsequent operators.
+    HAS_OUTPUT,
+};
+
+// TODO support operator profile info like `BlockStreamProfileInfo`.
+
+class Operator
+{
+public:
+    virtual ~Operator() = default;
+    // running status may return are
+    // - `NEED_INPUT` means that the data that the operator is waiting for has been prepared.
+    OperatorStatus await();
```

Comment:
It can also return `WAITING`, right?

Reply:
https://github.com/pingcap/tiflash/pull/6520/files#diff-eff9da86fd05cd3ebffddf28ca658334df6cf9ec43f8b56601f9e22e6d4f9cb3R24-R25
finish status, spilling status and waiting status can be returned in all interfaces of operator.